engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/StraightThroughProcessingLoopValidator.java

```diff
@@ -0,0 +1,222 @@
+   *
+   * @param resource the resource file that's getting validated
+   * @param executableProcess the process we are validating
+   * @return an Either of which the left contains the failure message if any loops have been found
+   */
+  private static Either<Failure, ?> hasStraightThroughProcessingLoop(
+      final DeploymentResource resource, final ExecutableProcess executableProcess) {
+    final var straightThroughElements = getStraightThroughElementsInProcess(executableProcess);
+
+    for (final ExecutableFlowNode element : straightThroughElements) {
+      final var potentialLoop = new LinkedList<ExecutableFlowNode>();
+      final var result = checkForStraightThroughProcessingLoop(potentialLoop, element);
+      if (result.isLeft()) {
+        final String failureMessage =
+            createFailureMessage(resource, executableProcess, result.getLeft());
+        return Either.left(new Failure(failureMessage));
+      }
+    }
+
+    return Either.right(null);
+  }
+
+  /**
+   * When looking for straight-through processing loops it is not necessary to check in all places
+   * of the process. We only have to check the path of straight-through elements. This method
+   * returns a list of all the straight-through elements in a process.
+   *
+   * @param executableProcess the process
+   * @return a list of all straight-through elements in the given process
+   */
+  private static List<ExecutableFlowNode> getStraightThroughElementsInProcess(
+      final ExecutableProcess executableProcess) {
+
+    return executableProcess.getFlowElements().stream()
+        .filter(
+            flowElement ->
+                STRAIGHT_THROUGH_PROCESSING_ELEMENT_TYPES.contains(flowElement.getElementType())
+                    && !WHITELISTED_ELEMENT_TYPES.contains(flowElement.getElementType()))
+        .map(ExecutableFlowNode.class::cast)
+        .toList();
+  }
+
+  /**
+   * This method is the part that actually tries to find any loops. It takes a LinkedList in which
+   * it keeps track of elements that are part of the (potential) loop. It also takes an element
+   * which is used to detect if we are in a loop.
+   *
+   * <p>Upon entering this method there are 3 possibilities:
+   *
+   * <ul>
+   *   <li>This element is part of the potentialLoop list. This means that we have passed this
+   *       element before. As a result we can conclude that we are in a loop.
+   *   <li>The element is not part of the potentialLoop list. We haven't passed the element before
+   *       so we are not in a loop yet. However, the element is a straight-through processing
+   *       element. We must check the outgoing sequence flow of this element to see if it loops back
+   *       around to an element that is part of the potentialLoop list.
+   *   <li>We have not encountered this element before, nor is it a straight-through processing
+   *       element. We can conclude that this path is not part of a straight-through processing
+   *       loop.
```

Comment:
> nor is it a straight-through processing

Not sure whether this can happen since it is filtered out right?

Reply:
I think there might be a slight understanding. We start with this filter:
```java
  private static List<ExecutableFlowNode> getStraightThroughElementsInProcess(
      final ExecutableProcess executableProcess) {

    return executableProcess.getFlowElements().stream()
        .filter(
            flowElement ->
                STRAIGHT_THROUGH_PROCESSING_ELEMENT_TYPES.contains(flowElement.getElementType())
                    && !WHITELISTED_ELEMENT_TYPES.contains(flowElement.getElementType()))
        .map(ExecutableFlowNode.class::cast)
        .toList();
  }
  ```

The elements returned here are the elements at which we **start** checking for a loop. This means if you have a process without manual/undefined tasks we don't even bother checking any loops.
Once we start checking we will take all outgoing flows of the elements returned by this filter. These outgoing flows can target any other element. If the outgoing flow targets a service task we know we are not in a straight-through loop. That's what I tried to convey with this comment.